packages/core/src/config.ts

```diff
@@ -955,64 +959,113 @@ const composeEntryConfig = async (
-          // Prevent from externalizing entry modules here.
-          if (data.contextInfo.issuer) {
-            // Node.js ECMAScript module loader does no extension searching.
-            // Add a file extension according to autoExtension config
-            // when data.request is a relative path and do not have an extension.
-            // If data.request already have an extension, we replace it with new extension
-            // This may result in a change in semantics,
-            // user should use copy to keep origin file or use another separate entry to deal this
-            let request: string = data.request;
-
-            const cssExternal = cssExternalHandler(
-              request,
-              callback,
-              jsExtension,
-              cssModulesAuto,
-              isStyleRedirect,
-            );
-
-            if (cssExternal !== false) {
-              return cssExternal;
+    resolvedJsRedirect: {
+      path: jsRedirectPath,
+      extension: jsRedirectExtension,
+    },
+    config: {
+      output: {
+        externals: [
+          async (data, callback) => {
+            const { request, getResolve, context, contextInfo } = data;
+            if (!request || !getResolve || !context || !contextInfo) {
+              return callback();
+            }
+
+            if (!resolver) {
+              resolver = (await getResolve()) as Resolver;
             }
 
-            if (request[0] === '.') {
-              const ext = extname(request);
+            // Issuer is not empty string when the module is imported by another module.
+            // Prevent from externalizing entry modules here.
+            if (contextInfo.issuer) {
+              let resolvedRequest: string = request;
+
+              const cssExternal = cssExternalHandler(
+                resolvedRequest,
+                callback,
+                jsExtension,
+                cssModulesAuto,
+                doesRedirectStyle,
+              );
+
+              if (cssExternal !== false) {
+                return cssExternal;
+              }
+
+              if (jsRedirectPath) {
+                try {
+                  resolvedRequest = await resolver(context, resolvedRequest);
+                } catch (e) {
+                  // Do nothing, fallthrough to other external matches.
+                  logger.debug(
+                    `Failed to resolve ${resolvedRequest} with resolver`,
+                  );
+                }
 
-              if (ext) {
-                if (JS_EXTENSIONS_PATTERN.test(request)) {
-                  request = request.replace(/\.[^.]+$/, jsExtension);
+                resolvedRequest = normalizeSlash(
+                  path.relative(
+                    path.dirname(contextInfo.issuer),
+                    resolvedRequest,
+                  ),
+                );
+
+                if (resolvedRequest[0] !== '.') {
```

Comment:
Why to convert to relative path？

Reply:
If a request is falling to here, it means it doesn't match any of the external rule. Which means it fall through the first three externals config to the `bundlelessExternalConfig`. Which will be treated as source code.

- autoExternalConfig,
- targetExternalsConfig,
- userExternalsConfig,
- bundlelessExternalConfig,

Reply:
I mean why adding `./` before `resolvedRequest` which is not begon with `.`

Reply:
It is, https://nodejs.org/api/path.html#pathrelativefrom-to does not add the leading `./`.